[Merged by Bors] - feat(category_theory/monad/*): Add category of bundled (co)monads; prove equivalence of monads with monoid objects

adamtopaz opened the pull request:
This PR constructs bundled monads, and proves the "usual" equivalence between the category of monads and the category of monoid objects in the endomorphism category.

It also includes a definition of morphisms of unbundled monads, and proves some necessary small lemmas in the following two files:
1. `category_theory.functor_category`
2. `category_theory.monoidal.internal`

Given any isomorphism in `Cat`, we construct a corresponding equivalence of categories in `Cat.equiv_of_iso`.

---
<!-- put comments you want to keep out of the PR commit here -->

adamtopaz committed:
Add bundled.lean, monad_hom, etc.

adamtopaz committed:
Various things:

1. Add hcomp_id_app / id_hcomp_app
2. Add Monad.hext
3. Add Mon_.hext
4. Add file equiv_mon.lean

adamtopaz committed:
Add eq_to_hom lemmas

adamtopaz committed:
Add Monad_Mon_equiv

adamtopaz committed:
A bit of cleanup.

adamtopaz committed:
More cleanup.

adamtopaz committed:
Add documentation

adamtopaz committed:
Cleanup

adamtopaz committed:
Cleanup

adamtopaz committed:
Add Cat.equiv_of_iso and rewrite Monad_Mon_equiv

b-mehta commented on src/category_theory/functor_category.lean:
```diff
@@ -65,6 +65,14 @@ infix ` ◫ `:80 := hcomp
+@[simp] lemma hcomp_id_app {H : D ⥤ E} (α : F ⟶ G) (X : C) : (α ◫ 𝟙 H).app X = H.map (α.app X) :=
+begin
+  simp only [id_app, hcomp_app],
+  apply id_comp,
```
Does it work to put `id_comp` in the `simp` arguments here?

adamtopaz replied:
This doesn't work for some reason.

b-mehta replied:
Frustrating, fair enough!

kim-em replied:
```
by { dsimp, simp, } -- See note [dsimp, simp].
```
suffices.

adamtopaz replied:
Done!

b-mehta commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,51 @@ attribute [simp] comonad.left_counit comonad.right_counit
 notation `δ_` := comonad.δ
 
+/-- A morphisms of monads is a natural transformation compatible with η and μ. -/
+structure monad_hom (M N : C ⥤ C) [monad M] [monad N] extends nat_trans M N :=
```
Would you mind dualising the things you've added here?

adamtopaz replied:
Done! (Also added bundled comonads  in category_theory.monad.bundled, and the corresponding category instance.)

b-mehta commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,51 @@ attribute [simp] comonad.left_counit comonad.right_counit
+
+restate_axiom monad_hom.app_η'
+restate_axiom monad_hom.app_μ'
+attribute [simp] monad_hom.app_η monad_hom.app_μ
```
I think it's worth marking these as `reassoc` as well

adamtopaz replied:
Done.

b-mehta commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,187 @@
+  inv_hom_id' := by apply of_to_mon_end }
+
+/-- Oh, monads are just monoids in the category of endofunctors (equivalence of categories). -/
+def Monad_Mon_equiv : equivalence (Monad C) (Mon_ (C ⥤ C)) := Cat.equiv_of_iso $ Monad_Mon_iso C
```
Is there a reason you avoided the notation for equivalence and isomorphism in these two?

adamtopaz replied:
Fixed (I just learned about \backcong)

adamtopaz committed:
Add morphism/categories for (bundled) comonds

adamtopaz committed:
Fix iso/equivalence notation

adamtopaz committed:
Fix docstring

adamtopaz committed:
Small cleanup

kim-em commented on src/category_theory/category/Cat.lean:
```diff
@@ -49,6 +50,13 @@ def objects : Cat.{v u} ⥤ Type u :=
+  inverse := γ.inv,
+  unit_iso := eq_to_iso $ eq.symm γ.hom_inv_id',
+  counit_iso := eq_to_iso γ.inv_hom_id' }
+
```
There's a missing coercion from `Cat` to `Type` in this file.

Can I suggest replacing this whole namespace block with:
```
namespace Cat

instance : inhabited Cat := ⟨⟨Type u, category_theory.types⟩⟩

instance : has_coe_to_sort Cat :=
{ S := Type u,
  coe := bundled.α }

instance str (C : Cat.{v u}) : category.{v u} C := C.str

/-- Construct a bundled `Cat` from the underlying type and the typeclass. -/
def of (C : Type u) [category.{v} C] : Cat.{v u} := bundled.of C

/-- Category structure on `Cat` -/
instance category : large_category.{max v u} Cat.{v u} :=
{ hom := λ C D, C ⥤ D,
  id := λ C, 𝟭 C,
  comp := λ C D E F G, F ⋙ G,
  id_comp' := λ C D F, by cases F; refl,
  comp_id' := λ C D F, by cases F; refl,
  assoc' := by intros; refl }

/-- Functor that gets the set of objects of a category. It is not
called `forget`, because it is not a faithful functor. -/
def objects : Cat.{v u} ⥤ Type u :=
{ obj := λ C, C,
  map := λ C D F, F.obj }

/-- Any isomorphism in `Cat` induces an equivalence of the underlying categories. -/
def equiv_of_iso {C D : Cat} (γ : C ≅ D) : C ≌ D :=
{ functor := γ.hom,
  inverse := γ.inv,
  unit_iso := eq_to_iso $ eq.symm γ.hom_inv_id,
  counit_iso := eq_to_iso γ.inv_hom_id }

end Cat
```

adamtopaz replied:
Done!

kim-em replied:
Ah, you didn't actually use my whole suggestion. I changed all the definitions to use the coercion rather than the `.α` projections. This isn't a big deal, but will potentially get in the way of syntactic matching further down the line. Better to write everything the same way. (We should possibly even add a simp lemma rewriting C.α as the coercion.)

adamtopaz replied:
Oops! I'll change this later today when I get a chance.

adamtopaz replied:
Okay, I think I got them all.

kim-em commented on src/category_theory/category/Cat.lean:
```diff
@@ -49,6 +50,13 @@ def objects : Cat.{v u} ⥤ Type u :=
+{ functor := γ.hom,
+  inverse := γ.inv,
+  unit_iso := eq_to_iso $ eq.symm γ.hom_inv_id',
+  counit_iso := eq_to_iso γ.inv_hom_id' }
```
```suggestion
  unit_iso := eq_to_iso $ eq.symm γ.hom_inv_id,
  counit_iso := eq_to_iso γ.inv_hom_id }
```

adamtopaz replied:
Done!

kim-em commented on src/category_theory/category/Cat.lean:
```diff
@@ -49,6 +50,13 @@ def objects : Cat.{v u} ⥤ Type u :=
+{ functor := γ.hom,
+  inverse := γ.inv,
+  unit_iso := eq_to_iso $ eq.symm γ.hom_inv_id',
+  counit_iso := eq_to_iso γ.inv_hom_id' }
```
I'm still worried this isn't the right definition, but failing to articulate my objection exactly. I worry that it might be better to "push the proofs down as far as possible". That is, use `nat_iso.of_components` to construct `unit_iso` and `counit.iso`, and specify that the components are given by `eq_to_iso` applied to `γ.hom_inv_id` "restricted" to a particular object.

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,98 @@ attribute [simp] comonad.left_counit comonad.right_counit
+  by {ext, apply comp_id}
+lemma comp_assoc (f : monad_hom M N) (g : monad_hom N L) (h : monad_hom L K) :
+  (f.comp g).comp h = f.comp (g.comp h) := by {ext, apply assoc}
+
```
Maybe leave a comment here pointing to the `bundled` file for the category instance?

adamtopaz replied:
Done!

kim-em commented on src/category_theory/monad/bundled.lean:
```diff
@@ -0,0 +1,144 @@
+end
+
+theorem to_nat_trans_eq_to_hom (M N : Monad C) (h : M = N) :
+  monad_hom.to_nat_trans (eq_to_hom h) = eq_to_hom (congr_arg (Monad.func) h) := by {subst h, refl}
```
Why do we want this? It feels yucky. :-)

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,187 @@
+
+/-- Oh, monads are just monoids in the category of endofunctors (equivalence of categories). -/
+def Monad_Mon_equiv : (Monad C) ≌ (Mon_ (C ⥤ C)) := Cat.equiv_of_iso $ Monad_Mon_iso C
+
```
Can we make

```
example (A : Monad C) {X : C} : ((Monad_Mon_equiv C).unit_iso.app A).hom.app X = 𝟙 (A.func.obj X) := rfl
```

work? At the moment it doesn't, I *think* because of the way you've sent up `Cat.equiv_of_iso`. But I'm worried I'm barking up the wrong tree here, so take this with a grain of salt.

adamtopaz committed:
A few changes:

1. Clean up proof of hcomp_id_app
2. Add has_coe_to_sort instance for Cat
3. Add note about bundled (co)monad category

adamtopaz committed:
Rework Monad_Mon_equiv

adamtopaz committed:
Add docstring

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+  { app := λ M,
+  { app := λ X, 𝟙 _,
+    app_η' := by tidy,
+    app_μ' := by tidy } },
```
You can probably omit these as the `auto_param` will handle them.

adamtopaz replied:
I had the same thought originally, but it didn't work for some reason. Did I declare the `auto_param` incorrectly in the definition of `monad_hom`?

adamtopaz replied:
Okay, I figured it out. When defining monad_hom I had fields of the form
`(app_η' {X} : ... . obviously)`
I changed it to 
`(app_η' : ∀ {X}, ... . obviously)`
and now the `auto_param` works as expected. Is this expected behavior?

kim-em replied:
No, I hadn't known about this issue. Can you distill it down to a minimal example? It seems worth reporting as an issue.

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+    ext A,
+    simp only [nat_trans.hcomp_id_app, nat_trans.hcomp_app, functor.map_id,
+      nat_trans.id_app, comp_id, nat_trans.comp_app],
+    erw id_comp,
```
```suggestion
    dsimp, 
    simp only [nat_trans.hcomp_id_app, nat_trans.hcomp_app, functor.map_id,
      nat_trans.id_app, comp_id, nat_trans.comp_app, id_comp],
```

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+      nat_trans.id_app, comp_id, nat_trans.comp_app],
+    erw id_comp,
+    simp_rw monad.assoc,
+    change _ = ((α_ M.func M.func M.func).app A).hom ≫ _ ≫ _,
```
Can these `change` steps be achieved with better simp lemmas?

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+local attribute [instance] endofunctor_monoidal_category
+
+/-- To every `Monad C` we associated a monoid object in `C ⥤ C`.-/
+def to_Mon : Monad C → Mon_ (C ⥤ C) := λ M,
```
You probably want to put `@[simps]` here to obtain all the definitional projection lemmas, and then in later proofs try to rely more on `simp`, and less on finishing steps with `apply` (which is requiring Lean to go think about the definitions, rather than just doing syntactic rewriting).

Often putting the work in the make `simp` work right is worth it, because it makes life easier for everyone who comes afterwards.

adamtopaz replied:
Done.

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+    one_hom' := begin
+      ext,
+      simp only [nat_trans.comp_app],
+      apply f.app_η,
```
This is a good example of something that I'd prefer worked by `simp`.

kim-em replied:
(and in fact could even then be omitted because an `auto_param` would handle it)

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+variable {C}
+
+theorem of_to_mon_end_obj (M : Mon_ (C ⥤ C)) : (Mon_to_Monad C ⋙ Monad_to_Mon C).obj M = M :=
+  by {apply Mon_.hext, repeat {refl}}
```
```suggestion
by {apply Mon_.hext, repeat {refl}}
```
No need to indent proofs/definitions. (Just indent multiline statements, and inside `begin ... end` blocks.)

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,233 @@
+  inv_hom_id' := by apply of_to_mon_end }
+
+/-- Oh, monads are just monoids in the category of endofunctors (equivalence of categories). -/
+def Monad_Mon_equiv : (Monad C) ≌ (Mon_ (C ⥤ C)) :=
```
```suggestion
@[simps]
def Monad_Mon_equiv : (Monad C) ≌ (Mon_ (C ⥤ C)) :=
```

adamtopaz replied:
Done. I also added `@[simps]` to the other definitions in this file.

adamtopaz committed:
Update src/category_theory/monad/equiv_mon.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

adamtopaz committed:
Helping tidy

adamtopaz committed:
A lot of cleanup

adamtopaz committed:
More cleanup

adamtopaz committed:
Small cleanup

adamtopaz committed:
Indentation fix

adamtopaz committed:
Add coassoc lemma

b-mehta commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+
+/-- The composition of two morphisms of comonads. -/
+def comp (f : comonad_hom M N) (g : comonad_hom N L) : comonad_hom M L :=
+{ app := λ X, (f.app X) ≫ (g.app X),
```
Redundant parens here

adamtopaz replied:
Thanks. Fixed.

b-mehta commented on src/category_theory/monad/bundled.lean:
```diff
@@ -0,0 +1,144 @@
+
+end Monad
+
+namespace CoMonad
```
I'm not a huge fan of `CoMonad`, I'd prefer `Comonad`, but this is definitely just my personal taste

adamtopaz replied:
I don't really have any strong opinions about this, although I do pronounce "comonad" as "co-monad" so "CoMonad" makes sense to me.

kim-em replied:
I think I'm with @b-mehta here, although there's little point bikeshedding it as Lean4 wants us all to change to CamelCase, so we have some renaming in the future anyway!

adamtopaz replied:
Okay, I renamed "CoMonad" to "Comonad".

adamtopaz committed:
Remove redundant parens

kim-em commented on src/category_theory/monoidal/internal.lean:
```diff
@@ -39,6 +39,16 @@ attribute [simp, reassoc] Mon_.one_mul Mon_.mul_one Mon_.mul_assoc
 
 namespace Mon_
 
+theorem hext (M N : Mon_ C) : M.X = N.X → M.one == N.one → M.mul == N.mul → M = N :=
```
Is this still used? Perhaps we could remove it as "evil" if not.

adamtopaz replied:
Yeah, I use this and the Monad analogue in `of_to_mon_end_obj` and `to_of_mon_end_obj`. I think these theorems could be useful. What do you think?

adamtopaz replied:
I just pushed a small change to these two theorems, which I think was necessary for them to be tagged as `simp`.

kim-em replied:
Sounds good!

kim-em replied:
Actually, now I think that `of_to_mon_end_obj` and `to_of_mon_end_obj` may not be fit for purpose. You don't use them in this PR, and as they are not definitional equalities it's going to be hard to use them --- you can't use them with `dsimp`, and `rw` and `simp` will fail to apply these equalities when you're looking at morphisms, because they'll break the dependent types.

So perhaps delete them, and this `hext`?

adamtopaz replied:
@semorrison Fair enough. I'll go ahead and delete these theorems, and the `hext` theorems for `Mon_` and `Monoid` before merging.

adamtopaz committed:
Change statemes of two theorems:
1. of_to_mon_end_obj
2. to_of_mon_end_obj

adamtopaz committed:
Cleanup using has_coe_to_sort

adamtopaz committed:
CoMonad -> Comonad

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+def ident : comonad_hom M M :=
+{ app_ε' := by tidy,
+  app_δ' := by tidy,
+  ..𝟙 M }
```
Could we make the naming scheme more uniform? Use just `id` here to match the rest of mathlib?

adamtopaz replied:
Done! Also renamed `ident_comp` as `id_comp`, etc.

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+  by {ext, apply comp_id}
+/-- Note: `category_theory.monad.bundled` provides a category instance for bundled comonads.-/
+lemma comp_assoc (f : comonad_hom M N) (g : comonad_hom N L) (h : comonad_hom L K) :
+  (f.comp g).comp h = f.comp (g.comp h) := by {ext, apply assoc}
```
Most of the category theory library uses just `assoc` instead of `comp_assoc`.

adamtopaz replied:
Done. Also added simp tags for these.

kim-em commented on src/category_theory/monad/bundled.lean:
```diff
@@ -0,0 +1,144 @@
+  comp := λ _ _ _, monad_hom.comp,
+  id_comp' := λ _ _, by apply monad_hom.ident_comp,
+  comp_id' := λ _ _, by apply monad_hom.comp_ident,
+  assoc' := λ _ _ _ _, by apply monad_hom.comp_assoc }
```
```suggestion
  id_comp' := λ _ _, monad_hom.ident_comp,
  comp_id' := λ _ _, monad_hom.comp_ident,
  assoc' := λ _ _ _ _, monad_hom.comp_assoc }
```

kim-em replied:
Or just leave out these fields entirely!

adamtopaz replied:
Done. I'm just relying on the auto-param now.

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+def ident : monad_hom M M :=
+{ app_η' := by tidy,
+  app_μ' := by tidy,
+  ..𝟙 M }
```
```suggestion
{ ..𝟙 M }
```

adamtopaz replied:
Done. I actually tried this before, but it didn't work, I think because of the same `{x} : ...` vs. ` ... : \forall {x}, ...` issue.

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+def comp (f : monad_hom M N) (g : monad_hom N L) : monad_hom M L :=
+{ app := λ X, f.app X ≫ g.app X,
+  app_η' := λ X, by {rw ←assoc, simp [app_η']},
+  app_μ' := λ X, by {rw ←assoc, simp [app_μ']} }
```
```suggestion
{ app := λ X, f.app X ≫ g.app X, }
```

adamtopaz replied:
Done.

kim-em commented on src/category_theory/monad/basic.lean:
```diff
@@ -56,4 +56,100 @@ attribute [simp] comonad.left_counit comonad.right_counit
+  app_μ' := λ X, by {rw ←assoc, simp [app_μ']} }
+
+@[simp] lemma ident_comp (f : monad_hom M N) : (monad_hom.ident M).comp f = f :=
+  by {ext, apply id_comp}
```
No indent needed after `:=`.

adamtopaz replied:
Done.

adamtopaz committed:
Renaming + cleanup

adamtopaz committed:
More cleanup

adamtopaz committed:
Fix indentation

adamtopaz committed:
Clean up spacing

adamtopaz committed:
Add simp tags on assoc, rely on auto-params

kim-em commented on src/category_theory/monad/equiv_mon.lean:
```diff
@@ -0,0 +1,134 @@
+  { hom := 𝟙 _ } },
+  inv :=
+  { app := λ M,
+  { hom := 𝟙 _ } } }
```
Not necessary at all, but if you prefer to make these more compact as 
```suggestion
{ hom := { app := λ M, { hom := 𝟙 _ } },
  inv := { app := λ M, { hom := 𝟙 _ } } }
```
I think that would be fine.

kim-em replied:
(I think the style guide is silent on this, but elsewhere lots of "single field structures" get put on the same line.)

adamtopaz replied:
I like the more compact style. I made an analogous change to `to_of_mon_end_iso` and `Monad_to_Mon`.

adamtopaz committed:
Update src/category_theory/monad/equiv_mon.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

adamtopaz committed:
Delete to_of/of_to_mon_end and hext lemmas

adamtopaz committed:
Final bit of cleanup

adamtopaz committed:
Final bit of cleanup
